File: src/main/java/net/adventuregame/hud/HudShader.java
package net.adventuregame.hud;

import com.chidozie.core.shaders.ShaderProgram;
import org.joml.Matrix4f;
import org.joml.Vector2f;

public class HudShader extends ShaderProgram {

    private static final String VERTEX_FILE = "hud/hudVertexShader.vert";
    private static final String FRAGMENT_FILE = "hud/hudFragmentShader.frag";

    private int location_mvpMatrix;
    private int location_useFakeLighting;
    private int location_numberOfRows;
    private int location_offset;
    private int location_isSelected;

    public HudShader() {
        super(VERTEX_FILE, FRAGMENT_FILE);
    }

    @Override
    protected void getAllUniformLocations() {
        location_mvpMatrix = getUniformLocation("mvpMatrix");
        location_useFakeLighting = getUniformLocation("useFakeLighting");
        location_numberOfRows = getUniformLocation("numberOfRows");
        location_offset = getUniformLocation("offset");
        location_isSelected = getUniformLocation("isSelected");
    }

    @Override
    protected void bindAttributes() {
        bindAttribute(0, "position");
        bindAttribute(1, "textureCoords");
        bindAttribute(2, "normal");
    }

    public void loadMvpMatrix(Matrix4f matrix) {
        loadMatrix(location_mvpMatrix, matrix);
    }

    public void loadUseFakeLighting(boolean useFake) {
        loadBoolean(location_useFakeLighting, useFake);
    }

    public void loadNumberOfRows(float rows) {
        loadFloat(location_numberOfRows, rows);
    }

    public void loadOffset(float x, float y) {
        loadVector2f(location_offset, new Vector2f(x, y));
    }

    public void loadIsSelected(boolean isSelected) {
        loadBoolean(location_isSelected, isSelected);
    }
}

File: src/main/kotlin/net/adventuregame/hud/HudShader.kt
package net.adventuregame.hud

import com.adv.core.shaders.ShaderProgram
import org.joml.Matrix4f
import org.joml.Vector2f

class HudShader : ShaderProgram(VERTEX_FILE, FRAGMENT_FILE) {
    private var location_mvpMatrix = 0
    private var location_useFakeLighting = 0
    private var location_numberOfRows = 0
    private var location_offset = 0
    private var location_isSelected = 0

    override fun getAllUniformLocations() {
        location_mvpMatrix = getUniformLocation("mvpMatrix")
        location_useFakeLighting = getUniformLocation("useFakeLighting")
        location_numberOfRows = getUniformLocation("numberOfRows")
        location_offset = getUniformLocation("offset")
        location_isSelected = getUniformLocation("isSelected")
    }

    override fun bindAttributes() {
        bindAttribute(0, "position")
        bindAttribute(1, "textureCoords")
        bindAttribute(2, "normal")
    }

    fun loadMvpMatrix(matrix: Matrix4f) {
        loadMatrix(location_mvpMatrix, matrix)
    }

    fun loadUseFakeLighting(useFake: Boolean) {
        loadBoolean(location_useFakeLighting, useFake)
    }

    fun loadNumberOfRows(rows: Float) {
        loadFloat(location_numberOfRows, rows)
    }

    fun loadOffset(x: Float, y: Float) {
        loadVector2f(location_offset, Vector2f(x, y))
    }

    fun loadIsSelected(isSelected: Boolean) {
        loadBoolean(location_isSelected, isSelected)
    }

    companion object {
        private const val VERTEX_FILE = "hud/hudVertexShader.vert"
        private const val FRAGMENT_FILE = "hud/hudFragmentShader.frag"
    }
}

